aten/src/ATen/native/TensorIndexing.h

```diff
@@ -77,36 +125,454 @@ struct CAFFE2_API TensorIndex final {
+  TensorIndex(std::initializer_list<c10::optional<int64_t>> slice) : type_(TensorIndexType::Slice) {
+    if (slice.size() == 0) {
+      slice_ = impl::unpackSlice(c10::nullopt, c10::nullopt, c10::nullopt);
+    } else if (slice.size() == 2) {
+      slice_ = impl::unpackSlice(*slice.begin(), *(slice.begin() + 1), c10::nullopt);
+    } else if (slice.size() == 3) {
+      slice_ = impl::unpackSlice(*slice.begin(), *(slice.begin() + 1), *(slice.begin() + 2));
+    } else {
+      TORCH_CHECK_VALUE(
+        false,
+        "Expected 0 / 2 / 3 elements in the braced-init-list to represent a slice index, but got ",
+        slice.size(),
+        " element(s)");
+    }
+  }
 
   // Case 5: Tensor value
-  TensorIndex(Tensor tensor);
+  TensorIndex(Tensor tensor) : tensor_(tensor), type_(TensorIndexType::Tensor) {}
+
+  inline bool is_none() const {
+    return type_ == TensorIndexType::None;
+  }
+
+  inline bool is_ellipsis() const {
+    return type_ == TensorIndexType::Ellipsis;
+  }
 
-  bool is_none() const;
-  bool is_ellipsis() const;
+  inline bool is_integer() const {
+    return type_ == TensorIndexType::Integer;
+  }
 
-  bool is_integer() const;
-  int64_t integer() const;
+  inline int64_t integer() const {
+    return integer_;
+  }
 
-  bool is_boolean() const;
-  bool boolean() const;
+  inline bool is_boolean() const {
+    return type_ == TensorIndexType::Boolean;
+  }
 
-  bool is_slice() const;
-  const Slice& slice() const;
+  inline bool boolean() const {
+    return boolean_;
+  }
 
-  bool is_tensor() const;
-  const Tensor& tensor() const;
+  inline bool is_slice() const {
+    return type_ == TensorIndexType::Slice;
+  }
+
+  inline const impl::Slice& slice() const {
+    return slice_;
+  }
+
+  inline bool is_tensor() const {
+    return type_ == TensorIndexType::Tensor;
+  }
+
+  inline const Tensor& tensor() const {
+    return tensor_;
+  }
 
  private:
   int64_t integer_;
   bool boolean_;
-  Slice slice_;
+  impl::Slice slice_;
   Tensor tensor_;
   TensorIndexType type_;
 };
 
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const TensorIndex& tensor_index);
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& tensor_indices);
 
+namespace impl {
+static inline Tensor applySlice(
+    const Tensor& self,
+    int64_t dim,
+    int64_t start,
+    int64_t stop,
+    int64_t step,
+    bool ensure_view,
+    bool is_tracing,
+    const at::Device& self_device,
+    const IntArrayRef& self_sizes) {
+  // TODO: implement negative step
+  TORCH_CHECK_VALUE(step > 0, "step must be greater than zero");
+
+  // Skip this optimization if we are tracing, as the trace may be polymorphic
+  // over the shape of the `self` tensor, and we still want to record
+  // the slice.
+  int64_t length = (self_device == at::kCPU || self_device == at::kCUDA) ? self_sizes[dim] : self.size(dim);
+  if (!ensure_view && start == 0 && stop == length && step == 1 && !is_tracing) {
```

Comment:
This function has an irritatingly large number of parameters, so I looked into what was goin on with `ensure_view`. What I think is going on with `ensure_view`, is that it's a sort of trick to reduce the amount of code we have to write at down stream sites. The underlying motivation for `ensure_view` is to let you avoid writing code like this:

```
  Variable sliced = applySlicing(self_, holder.get(), variableIndices);
  if (variableIndices.empty()) {
    if (sliced.is_same(self_)) {
      // ensure we return a shallow copy for things like x[...]
      sliced = at::alias(sliced);
    }
    return wrap(sliced);
  }
```

for getters, as getters are guaranteed to always return a new view, no matter what (for setters, you can avoid returning a new view.) So the shorter version of this code looks like:

```
 else if (PySlice_Check(index)) {
    return wrap(applySlice(self_, 0, index, true));
  }
```

Here, we passed in `true` to say, "Hey, you always have to return a view here", and so we'll end up with a slice view here. BTW, we can't (easily?) do this in the applySlicing case, because the actual call to applySlice is done inside some sort of loop, where we only need to make sure an alias gets created at the very end (e.g., it is wasteful to create fresh aliases for each index in `x[:,:,:]`, we only need to do it at the end."

This original intent of the code got substantially garbled after your refactoring to pass in `is_tracing` as an argument, as we now have two boolean arguments which serve exactly the same purpose, but no clear semantic distinction between them. 

I'm not exactly sure what to do here (perhaps the booleans should be combined?), but it's worth thinking about. I started looking into this because I saw `is_tracing_and_1d_slice_or_Nd` which is rubbing me the wrong way.